dart/collision/CollisionGroup.h

```diff
@@ -0,0 +1,243 @@
+                           const Others*... others);
+
+  /// Remove ShapeFrames of Skeleton, and also remove another ShapeFrames of
+  /// other various objects.
+  template <typename... Others>
+  void removeShapeFramesOf(const dynamics::Skeleton* skeleton,
+                           const Others*... others);
+
+  /// Do nothing. This function is for terminating the recursive variadic
+  /// template function template<typename...> removeShapeFramesOf().
+  void removeShapeFramesOf();
+
+  /// Remove all the ShapeFrames in this CollisionGroup
+  void removeAllShapeFrames();
+
+  /// Return true if this CollisionGroup contains shapeFrame
+  bool hasShapeFrame(const dynamics::ShapeFrame* shapeFrame) const;
+
+  /// Return number of ShapeFrames added to this CollisionGroup
+  size_t getNumShapeFrames() const;
+
+  /// Perform collision detection within this CollisionGroup.
+  bool collide(const CollisionOption& option, CollisionResult& result);
+
+  /// Perform collision detection with other CollisionGroup.
+  ///
+  /// Return false if the engine type of the other CollisionGroup is different
+  /// from this CollisionObject engine.
+  bool collide(CollisionGroup* group, const CollisionOption& option, CollisionResult& result);
+
+protected:
+
+  /// Update engine data. This function should be called before the collision
+  /// detection is performed by the engine in most cases.
+  void updateEngineData();
+
+  /// Initialize the collision detection engine data such as broadphase
+  /// algorithm. This function will be called whenever ShapeFrame is either
+  /// added to or removed from this CollisionGroup.
+  virtual void initializeEngineData() = 0;
+
+  /// Add CollisionObject to the collision detection engine
+  virtual void addCollisionObjectToEngine(CollisionObject* object) = 0;
+
+  /// Add CollisionObjects to the collision detection engine
+  virtual void addCollisionObjectsToEngine(
+      const std::vector<CollisionObject*>& collObjects) = 0;
+
+  /// Remove CollisionObject from the collision detection engine
+  virtual void removeCollisionObjectFromEngine(CollisionObject* object) = 0;
+
+  /// Remove all the CollisionObjects from the collision detection engine
+  virtual void removeAllCollisionObjectsFromEngine() = 0;
+
+  /// Update the collision detection engine data such as broadphase algorithm.
+  /// This function will be called ahead of every collision checking.
+  virtual void updateCollisionGroupEngineData() = 0;
+
+protected:
+
+  /// Collision detector
+  CollisionDetectorPtr mCollisionDetector;
+  // CollisionGroup shares the ownership of CollisionDetector with other
+  // CollisionGroups created from the same CollisionDetector so that the
+  // CollisionDetector doesn't get destroyed as long as at least one
+  // CollisionGroup is alive.
+
+  /// ShapeFrames and CollisionOjbects added to this CollisionGroup
+  std::vector<std::pair<const dynamics::ShapeFrame*,
+                        CollisionObjectPtr>> mShapeFrameMap;
```

Comment:
Yeah, that approach sounds good. We could implement something similar to [`nlohmann/fifo_map`](https://github.com/nlohmann/fifo_map) for the sync guarantee.

I would like to do this as a separate PR since this PR blocks others.